olm/operator-registry.md

```diff
@@ -0,0 +1,318 @@
+
+In order to serve that manifest image, we cannot pull the image directly. Instead we will:
+
+1. Create a pod spec that uses an init container to inject a binary into the manifest image.
+2. Run the manifest image with that binary that writes to a configmap in a well known location
+3. Read the configmap data
+4. On return kill the pod
+5. Delete the configmap
+
+pod spec:
+```
+apiVersion: v1
+kind: Pod
+metadata:
+  name: example
+spec:
+  containers:
+  - name: operator-manifest
+    image: quay.io/community-operators/foo@sha256:abc123
+    command: ['operator-registry', 'serve' '-manifestdir "./manifests/"']
+  initContainers:
+  - name: init-manifest
+    image: quay.io/operator-framework/init-operator-manifest:latest
+    command: ['cp', 'operator-registry', '.']
+```
+
+We will also need to update operator-registry to include a new `serve` command that knows how to parse the manifest/ folder in the operator bundle image. This command will write to a configmap that OLM will use to get the bundle data.
+
+*DB Versioning*
+
+Since we are now calling this database schema a first class releasable object, we also need to be able to make changes to the schema over time. Along with that, we will need to be able to migrate to newer schema versions over time. To get started with that, we will update the existing operator-registry database to include a table that defines the current version of the database schema. At that point, the DB format is versioned with schemas.
+
+To support this, we will also add a version to the built operator-registry binary tooling (which currently exists in an unversioned state). Going forward, this tooling will support a set of different versions and handle migrations as needed to the latest schema that the tool's version supports. To go along with this, the runtime GRPC API that serves data to OLM will be versioned with the rest of the tooling (by the nature of the space it exists in) but will remain backwards compatible for API calls.
+
+Any large changes we wish to make (i.e. want to deprecate the current grpc api because we've changed how dependency resolution works) will be indicated via new CatalogSource types (current types are grpc, and configmap/internal). We will also provide a migration strategy along with any such changes.
+
+*Make CatalogSource type for this image*
+
+Finally, once these operator registry images are servable, we will update the catalog source to be able to download these images and serve them.
+
+Additionally, we will update the spec to create a polling field that allows the index to be updated automatically on a configurable poll timer.
+
+```
+apiVersion: operators.coreos.com/v1alpha1
+kind: CatalogSource
+metadata:
+  name: example-catalog
+  namespace: olm
+spec:
+  sourceType: grpc
```

Comment:
sourceType grpc? not "image" or something?

Reply:
This isn't a net new API, I am planning on just reusing the existing type. OLM already supports handling images that host grpc apis as input, so the intent here is to reuse the existing API that users are already familiar with and just expanding on it.